File: 2118-111-2115-find-all-possible-recipes-from-given-supplies/2118-111-2115-find-all-possible-recipes-from-given-supplies.js

var findAllRecipes = function(recipes, ingredients, supplies) {
    const adj = new Map();
    const inDegree = new Map();
    
    for (const recipe of recipes) {
        inDegree.set(recipe, 0);
    }
    
    for (let i = 0; i < recipes.length; i++) {
        for (const ingredient of ingredients[i]) {
            if (!adj.has(ingredient)) {
                adj.set(ingredient, []);
            }
            adj.get(ingredient).push(recipes[i]);
            inDegree.set(recipes[i], inDegree.get(recipes[i]) + 1);
        }
    }
    
    const ans = [];
    
    while (supplies.length > 0) {
        const node = supplies.pop();
        if (adj.has(node)) {
            for (const recipe of adj.get(node)) {
                inDegree.set(recipe, inDegree.get(recipe) - 1);
                if (inDegree.get(recipe) === 0) {
                    supplies.push(recipe);
                    ans.push(recipe);
                }
            }
        }
    }
    
    return ans;

};

File: 2118-111-2115-find-all-possible-recipes-from-given-supplies/2118-111-2115-find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Map<String, List<String>> adj = new HashMap<>();
        Map<String, Integer> inDegree = new HashMap<>();
        
        for (String recipe : recipes) {
            inDegree.put(recipe, 0);
        }
        
        for (int i = 0; i < recipes.length; i++) {
            for (String ingredient : ingredients.get(i)) {
                adj.putIfAbsent(ingredient, new ArrayList<>());
                adj.get(ingredient).add(recipes[i]);
                inDegree.put(recipes[i], inDegree.get(recipes[i]) + 1);
            }
        }
        
        List<String> ans = new ArrayList<>();
        List<String> suppliesList = new ArrayList<>(Arrays.asList(supplies));
        
        while (!suppliesList.isEmpty()) {
            String node = suppliesList.remove(suppliesList.size() - 1);
            if (adj.containsKey(node)) {
                for (String recipe : adj.get(node)) {
                    inDegree.put(recipe, inDegree.get(recipe) - 1);
                    if (inDegree.get(recipe) == 0) {
                        suppliesList.add(recipe);
                        ans.add(recipe);
                    }
                }
            }
        }
        
        return ans;
    }
    
}